### PR TITLE
Add documentation for CI pipelines

### PR DESCRIPTION
**Pull request by dkorpel**
I'm no expert on the CI by any means, so some explanations are incomplete and may be inaccurate. I hope the people who set up / maintain these checks can help fill in the details.

### CODE REVIEW

File: ci/README.md [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+If you don't have an `upstream` remote, you can add it with:
+```
+git remote add upstream git@github.com:dlang/dmd.git
+```
+
+Then, pull the latest changes from the branch that your PR merged into (`master` or `stable`) and rebase your branch.
+```
+git checkout master
+git pull --ff-only upstream master
+git checkout your-branch
```
Comment (maxhaton) [inline ci/README.md:30]: You can just `git fetch` then `git rebase dlang/master`, no?

File: ci/README.md [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+**Config**: [.circleci/config.yml](https://github.com/dlang/dmd/blob/master/.circleci/config.yml)
+
+**Checks**:
+- ci/circleci: build
+
+CircleCI tests DMD on Ubuntu 18.04.
+
+It also tests that the automatically generated C++ header, frontend.h, is updated.
+This is important because other compilers sharing the front-end (LDC and GDC) rely on DMD's header files to interface with it.
+When a PR modifies an `extern(C++)` function, the corresponding signature in a .h file should be updated as well, see [cxx-headers-test](https://github.com/dkorpel/dmd/tree/master/src#cxx-headers-test).
```
Comment (maxhaton) [inline ci/README.md:202]: This is a link to your fork not the main dmd repo.

File: ci/README.md [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+### DAutoTest
+
+**Config**: [CyberShadow/DAutoTest](https://github.com/CyberShadow/DAutoTest)
+
+**Checks**:
+- CyberShadow/DAutoTest — Documentation
+
+DMD includes a documentation generator, which is also used to build [the dlang.org website](https://github.com/dlang/dlang.org/).
+DAutoTest tests that the PR doesn't break the documentation generator, and also tests [changelog entries](https://github.com/dlang/dmd/tree/master/changelog) for correct DDoc syntax.
```
Comment (maxhaton) [inline ci/README.md:77]: Does it test the doc generator or test the generated docs? I thought the latter.

**Comment by maxhaton** [comment]
Probably worth explicitly noting that the LDC/GDC tests don't test the changes on top of ldc and gdc but rather compiling dmd with them.

File: ci/README.md (hunk 1 above)
Comment (dkorpel) [reply ci/README.md:30]: I guess this is just my [workaround](https://xkcd.com/763/)

File: ci/README.md (hunk 3 above)
Comment (dkorpel) [reply ci/README.md:77]: With 'test generated docs', do you mean test code examples in documentation? In any case, @CyberShadow can you describe what DAutoTest does exactly?

File: ci/README.md (hunk 2 above)
Comment (dkorpel) [reply ci/README.md]: Fixed

File: ci/README.md (hunk 3 above)
Comment (CyberShadow) [reply ci/README.md:74]: - Builds the documentation by invoking the dlang.org makefile.
  - This necessarily also builds the compiler, runtime, standard library, and fetches some dependencies.
- Saves the resulting HTML files and makes them available for preview.
- Allows looking at the differences in generated HTML between the base branch and the PR branch.
- Not DAutoTest but an adjacent component with the same underlying implementation also builds the master branch when it's updated (for any core repository), publishes the result on http://dlang.k3.1azy.net/, and uploads it to the official website.

Comment (dkorpel) [reply ci/README.md]: Thanks

File: ci/README.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+- macOS 10.15 x64, DMD (bootstrap)
+- macOS 11.x x64, DMD (coverage)
+- macOS 11.x x64, DMD (latest)
+- macOS 12.x x64, DMD (coverage)
+- macOS 12.x x64, DMD (latest)
+
+Cirrus tests DMD on Posix platforms.
+
+Since DMD is written in D, a "Host D Compiler" is needed to compile it.
+Various host compilers are tested, such as GDC, LDC, the latest DMD, and an older verison of DMD (bootstrap).
```
Comment (RazvanN7) [inline ci/README.md:163]:
```suggestion
Various host compilers are tested, such as GDC, LDC, the latest DMD, and an older version of DMD (bootstrap).
```

File: ci/README.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+Note that the GDC and LDC targets only test a different host compiler, they still compile the DMD compiler only.
```
Comment (RazvanN7) [inline ci/README.md:164]: This phrase is kind of ambiguous. I don't understand what you are trying to say.

File: ci/README.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+- pre_commit
+
+pre-commit checks for redundant white space, such as trailing whitespace and newlines.
+
+If you want to automatically make your changes comply with the check, you can download `pre-commit` via your package manager, or with  `pip`.
+See https://pre-commit.com/.
+
+Run `pre-commit install` inside the root of the repository,
+This will add a hook into `.git/hooks/pre-commit`.
+Every time you do a commit it will do pre-commit checks.
```
Comment (RazvanN7) [inline ci/README.md:183]:
```suggestion
Every time you make a commit it will do pre-commit checks.
```

File: ci/README.md [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+
+If you already have a custom `pre-commit` hook, you can modify the bash script to run `pre-commit` manually with `pre-commit run`.
+pre-commit is designed to run only on diffed files so it will run pretty quickly, but if you want to for it to run on all files you can do `pre-commit run --all-files`.
```
Comment (RazvanN7) [inline ci/README.md:186]:
```suggestion
pre-commit is designed to run only on diffed files so it will run pretty quickly, but if you want it to run on all files you can do `pre-commit run --all-files`.
```

File: ci/README.md [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+### CircleCI
+
+**Config**: [.circleci/config.yml](https://github.com/dlang/dmd/blob/master/.circleci/config.yml)
+
+**Checks**:
+- ci/circleci: build
+
+CircleCI tests DMD on Ubuntu 18.04.
+
+It also tests that the automatically generated C++ header, frontend.h, is updated.
```
Comment (RazvanN7) [inline ci/README.md:208]:
```suggestion
It also tests that the automatically generated C++ header, `frontend.h`, is updated.
```

**Review by RazvanN7 — CHANGES_REQUESTED** [review]

File: ci/README.md (hunk 5 above)
Comment (dkorpel) [reply ci/README.md:164]: Max said "Probably worth explicitly noting that the LDC/GDC tests don't test the changes on top of ldc and gdc but rather compiling dmd with them."

Comment (RazvanN7) [reply ci/README.md:164]: Then I would rephrase to: "Note that the GDC and LDC targets do not run any tests, the simply build the latest version of the dmd frontend."

Comment (dkorpel) [reply ci/README.md:164]: Don't they run the test suite?

File: ci/README.md [hunk 9]
```diff
@@ -0,0 +1,210 @@
+# Continuous Integration
+
+When a Pull Request (PR) is opened, several automated checks are performed to ensure the new compiler still builds, passes tests, and does not break external projects.
+Since DMD runs on many platforms and lots of D infrastructure depends on it, they are spread over many CI services.
+
+## Troubleshooting failed checks
+When your PR fails checks, click on the "Details" link next to it to see what the error is.
+
+Errors caused by the changes you made should be addressed by pushing new commits to your branch.
+
```
Comment (maxhaton) [inline ci/README.md:10]:
```suggestion
### How to rebase
```
This is a common issue so best to make it obvious.

**Review by RazvanN7 — APPROVED** [review]

File: ci/README.md [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+If your branch is already based on the latest commit, you get this message:
+```
+Current branch your-branch is up to date.
+```
+
+If you still want to re-run the checks, you can amend your last commit and force push:
+```
+git commit --amend
+git push origin master --force
```
[the remark this answers is not in the document]
Comment (dkorpel) [reply ci/README.md:51]: https://github.com/dlang/dmd/pull/14145
